File: app/src/main/java/de/cineaste/android/adapter/series/SeriesDetailAdapter.java
package de.cineaste.android.adapter.series;


import android.content.Context;
import android.content.res.Resources;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import de.cineaste.android.R;
import de.cineaste.android.entity.series.Season;
import de.cineaste.android.entity.series.Series;
import de.cineaste.android.listener.ItemClickListener;

public class SeriesDetailAdapter extends RecyclerView.Adapter {

    public interface SeriesStateManipulationClickListener {
        void onDeleteClicked();
        void onAddToHistoryClicked();
        void onAddToWatchClicked();
    }

    private Series series;
    private ItemClickListener clickListener;
    private int state;
    private SeriesStateManipulationClickListener listener;

    public SeriesDetailAdapter(Series series, ItemClickListener clickListener, int state, SeriesStateManipulationClickListener listener) {
        this.series = series;
        List<Season> seasons = new ArrayList<>();
        for (Season season : series.getSeasons()) {
            if (season.getSeasonNumber() > 0) {
                seasons.add(season);
            }
        }
        this.series.getSeasons().clear();
        this.series.setSeasons(seasons);
        this.clickListener = clickListener;
        this.state = state;
        this.listener = listener;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v;
        switch (viewType) {
            case 0:
                v = LayoutInflater.from(parent.getContext()).inflate(R.layout.series_detail_base, parent, false);
                return new BaseViewHolder(v, parent.getContext());
            case 1:
                v = LayoutInflater.from(parent.getContext()).inflate(R.layout.series_detail_buttons, parent, false);
                return new ButtonsViewHolder(v, state, listener);
            case 2:
                v = LayoutInflater.from(parent.getContext()).inflate(R.layout.series_detail_description, parent, false);
                return new DescriptionViewHolder(v, parent.getContext());
            case 3:
                v = LayoutInflater.from(parent.getContext()).inflate(R.layout.series_detail_seasons, parent, false);
                return new SeasonsListViewHolder(v, parent.getContext(), clickListener);
        }

        return null;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        switch (position) {
            case 0:
                ((BaseViewHolder) holder).assignData(series);
                break;
            case 1:
                ((ButtonsViewHolder) holder).assignData();
                break;
            case 2:
                ((DescriptionViewHolder) holder).assignData(series);
                break;
            case 3:
                ((SeasonsListViewHolder) holder).assignData(series);
                break;
        }
    }

    @Override
    public int getItemCount() {
//        if (state == R.string.searchState)
//            return 3;
        return 4;
    }

    @Override
    public int getItemViewType(int position) {
        return position;
    }

    private class BaseViewHolder extends RecyclerView.ViewHolder {
        private TextView rating;
        private TextView title;
        private TextView seasons;
        private TextView episodes;
        private TextView currentStatus;
        private TextView releaseDate;
        private TextView toBeContinued;
        private Resources resources;

        BaseViewHolder(View itemView, Context context) {
            super(itemView);
            this.rating = itemView.findViewById(R.id.rating);
            this.title = itemView.findViewById(R.id.title);
            this.seasons = itemView.findViewById(R.id.seasons);
            this.episodes = itemView.findViewById(R.id.episodes);
            this.currentStatus = itemView.findViewById(R.id.currentStatus);
            this.releaseDate = itemView.findViewById(R.id.releaseDate);
            this.toBeContinued = itemView.findViewById(R.id.toBeContinued);
            this.resources = context.getResources();
        }

        void assignData(Series series) {
            title.setText(series.getName());
            if (series.getReleaseDate() != null) {
                releaseDate.setText(convertDate(series.getReleaseDate()));
                releaseDate.setVisibility(View.VISIBLE);
            } else {
                releaseDate.setVisibility(View.GONE);
            }

            rating.setText(String.valueOf(series.getVoteAverage()));
            episodes.setText(resources.getString(R.string.episodes, String.valueOf(series.getNumberOfEpisodes())));
            seasons.setText(resources.getString(R.string.seasons, String.valueOf(series.getNumberOfSeasons())));
            currentStatus.setText(resources.getString(R.string.currentStatus,
                    String.valueOf(series.getCurrentNumberOfSeason()),
                    String.valueOf(series.getCurrentNumberOfEpisode())));
            if (series.isInProduction()) {
                toBeContinued.setVisibility(View.VISIBLE);
            } else {
                toBeContinued.setVisibility(View.GONE);
            }
        }

        private String convertDate(Date date) {
            SimpleDateFormat simpleDateFormat = new SimpleDateFormat("dd.MM.yyyy", resources.getConfiguration().locale);
            return simpleDateFormat.format(date);
        }

    }

    private class ButtonsViewHolder extends RecyclerView.ViewHolder {
        private Button deleteBtn;
        private Button historyBtn;
        private Button watchListBtn;
        private SeriesStateManipulationClickListener listener;

        ButtonsViewHolder(View itemView, int state, SeriesStateManipulationClickListener listener) {
            super(itemView);
            deleteBtn = itemView.findViewById(R.id.delete_button);
            historyBtn = itemView.findViewById(R.id.history_button);
            watchListBtn = itemView.findViewById(R.id.to_watchlist_button);
            this.listener = listener;

            switch (state) {
                case R.string.searchState:
                    deleteBtn.setVisibility(View.GONE);
                    historyBtn.setVisibility(View.VISIBLE);
                    watchListBtn.setVisibility(View.VISIBLE);
                    break;
                case R.string.historyState:
                    deleteBtn.setVisibility(View.VISIBLE);
                    historyBtn.setVisibility(View.GONE);
                    watchListBtn.setVisibility(View.VISIBLE);
                    break;
                case R.string.watchlistState:
                    deleteBtn.setVisibility(View.VISIBLE);
                    historyBtn.setVisibility(View.VISIBLE);
                    watchListBtn.setVisibility(View.GONE);
                    break;
            }
        }

        public void assignData() {
            deleteBtn.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    listener.onDeleteClicked();
                }
            });

            historyBtn.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    listener.onAddToHistoryClicked();
                }
            });

            watchListBtn.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    listener.onAddToWatchClicked();
                }
            });
        }
    }

    private class DescriptionViewHolder extends RecyclerView.ViewHolder {
        private TextView description;
        private TextView more;
        private Resources resources;

        DescriptionViewHolder(View itemView, Context context) {
            super(itemView);
            this.description = itemView.findViewById(R.id.description);
            this.more = itemView.findViewById(R.id.more);
            this.resources = context.getResources();
        }

        void assignData(final Series series) {
            more.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    showCompleteText(series);
                }
            });
            description.setText(getTrimmedDescription(series));
        }

        private void setDescription(Series series) {
            String original = series.getDescription();
            description.setText(
                    (original == null || original.isEmpty())
                            ? resources.getString(R.string.noDescription) : original);
        }

        private void showCompleteText(Series series) {
            setDescription(series);
            more.setVisibility(View.GONE);

        }

        private String getTrimmedDescription(Series series) {
            String original = series.getDescription();

            if (original == null || original.isEmpty()) {
                return resources.getString(R.string.noDescription);
            }

            if (original.length() <= 200) {
                more.setVisibility(View.GONE);
                return original;
            }
            more.setVisibility(View.VISIBLE);
            return original.substring(0, 200) + "...";
        }
    }

    private class SeasonsListViewHolder extends RecyclerView.ViewHolder {
        private RecyclerView recyclerView;
        private Context context;
        private ItemClickListener itemClickListener;

        SeasonsListViewHolder(View itemView, Context context, ItemClickListener itemClickListener) {
            super(itemView);

            this.recyclerView = itemView.findViewById(R.id.seasonPoster);
            this.context = context;
            this.itemClickListener = itemClickListener;
        }

        void assignData(Series series) {
            LinearLayoutManager layoutManager = new LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false);

            recyclerView.setLayoutManager(layoutManager);
            recyclerView.setItemAnimator(new DefaultItemAnimator());
            recyclerView.setNestedScrollingEnabled(false);

            SeasonAdapter adapter = new SeasonAdapter(context, itemClickListener, series);
            recyclerView.setAdapter(adapter);
        }
    }
}

File: app/src/main/kotlin/de/cineaste/android/adapter/series/SeriesDetailAdapter.kt
package de.cineaste.android.adapter.series

import android.content.Context
import android.content.res.Resources
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import com.squareup.picasso.Picasso

import java.text.SimpleDateFormat
import java.util.Locale
import java.util.Date

import de.cineaste.android.R
import de.cineaste.android.entity.series.Season
import de.cineaste.android.entity.series.Series
import de.cineaste.android.listener.ItemClickListener
import de.cineaste.android.util.Constants

class SeriesDetailAdapter(
    private var series: Series,
    private val clickListener: ItemClickListener,
    private val state: Int,
    private val listener: SeriesStateManipulationClickListener,
    private val posterClickListener: View.OnClickListener
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    interface SeriesStateManipulationClickListener {
        fun onDeleteClicked()
        fun onAddToHistoryClicked()
        fun onAddToWatchClicked()
    }

    init {
        val seasons = mutableListOf<Season>()
        for (season in series.seasons) {
            if (season.seasonNumber > 0) {
                seasons.add(season)
            }
        }
        this.series.seasons = seasons
    }

    fun updateSeries(series: Series) {
        this.series = series
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        when (viewType) {
            0 -> {
                val view = LayoutInflater
                    .from(parent.context)
                    .inflate(R.layout.series_detail_triangle, parent, false)
                return TriangleViewHolder(view)
            }
            1 -> {
                val view = LayoutInflater
                    .from(parent.context)
                    .inflate(R.layout.series_detail_base, parent, false)
                return BaseViewHolder(view, parent.context, posterClickListener)
            }
            2 -> {
                val view = LayoutInflater
                    .from(parent.context)
                    .inflate(R.layout.series_detail_buttons, parent, false)
                return ButtonsViewHolder(view, state, listener)
            }
            3 -> {
                val view = LayoutInflater
                    .from(parent.context)
                    .inflate(R.layout.series_detail_description, parent, false)
                return DescriptionViewHolder(view, parent.context)
            }
            4 -> {
                val view = LayoutInflater
                    .from(parent.context)
                    .inflate(R.layout.series_detail_seasons, parent, false)
                return SeasonsListViewHolder(view, parent.context, clickListener)
            }
            else -> {
                val view = LayoutInflater
                    .from(parent.context)
                    .inflate(R.layout.series_detail_base, parent, false)
                return BaseViewHolder(view, parent.context, posterClickListener)
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (position) {
            0 -> (holder as TriangleViewHolder).assignData(series)
            1 -> (holder as BaseViewHolder).assignData(series)
            2 -> (holder as ButtonsViewHolder).assignData()
            3 -> (holder as DescriptionViewHolder).assignData(series)
            4 -> (holder as SeasonsListViewHolder).assignData(series)
        }
    }

    override fun getItemCount(): Int {
        return 5
    }

    override fun getItemViewType(position: Int): Int {
        return position
    }

    private inner class TriangleViewHolder internal constructor(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
        private val rating: TextView = itemView.findViewById(R.id.rating)

        internal fun assignData(series: Series) {
            rating.text = series.voteAverage.toString()
        }
    }

    private inner class BaseViewHolder
    internal constructor(
        itemView: View,
        val context: Context,
        val clickListener: View.OnClickListener
    ) : RecyclerView.ViewHolder(itemView) {
        private val poster: ImageView = itemView.findViewById(R.id.poster)
        private val title: TextView = itemView.findViewById(R.id.title)
        private val seasons: TextView = itemView.findViewById(R.id.seasons)
        private val episodes: TextView = itemView.findViewById(R.id.episodes)
        private val currentStatus: TextView = itemView.findViewById(R.id.currentStatus)
        private val releaseDate: TextView = itemView.findViewById(R.id.releaseDate)
        private val toBeContinued: TextView = itemView.findViewById(R.id.toBeContinued)
        private val resources: Resources = context.resources

        internal fun assignData(series: Series) {
            title.text = series.name
            if (series.releaseDate != null) {
                releaseDate.text = convertDate(series.releaseDate)
                releaseDate.visibility = View.VISIBLE
            } else {
                releaseDate.visibility = View.GONE
            }

            episodes.text =
                resources.getString(R.string.episodes, series.numberOfEpisodes.toString())
            seasons.text = resources.getString(R.string.seasons, series.numberOfSeasons.toString())
            currentStatus.text = resources.getString(
                R.string.currentStatus,
                series.currentNumberOfSeason.toString(),
                series.currentNumberOfEpisode.toString()
            )
            if (series.isInProduction) {
                toBeContinued.visibility = View.VISIBLE
            } else {
                toBeContinued.visibility = View.GONE
            }

            setPoster(series)
            poster.setOnClickListener(clickListener)
        }

        private fun setPoster(series: Series) {
            val posterName = series.posterPath
            val posterUri = Constants.POSTER_URI_SMALL
                .replace("<posterName>", posterName ?: "/")
                .replace("<API_KEY>", context.getString(R.string.movieKey))
            Picasso.get()
                .load(posterUri)
                .resize(273, 410)
                .error(R.drawable.placeholder_poster)
                .into(poster)
        }

        private fun convertDate(date: Date?): String {
            val simpleDateFormat = SimpleDateFormat("dd.MM.yyyy", Locale.ENGLISH)
            return simpleDateFormat.format(date)
        }
    }

    private inner class ButtonsViewHolder
    internal constructor(
        itemView: View,
        state: Int,
        private val listener: SeriesStateManipulationClickListener
    ) : RecyclerView.ViewHolder(itemView) {
        private val deleteBtn: Button = itemView.findViewById(R.id.delete_button)
        private val historyBtn: Button = itemView.findViewById(R.id.history_button)
        private val watchListBtn: Button = itemView.findViewById(R.id.to_watchlist_button)

        init {

            when (state) {
                R.string.searchState -> {
                    deleteBtn.visibility = View.GONE
                    historyBtn.visibility = View.VISIBLE
                    watchListBtn.visibility = View.VISIBLE
                }
                R.string.historyState -> {
                    deleteBtn.visibility = View.VISIBLE
                    historyBtn.visibility = View.GONE
                    watchListBtn.visibility = View.VISIBLE
                }
                R.string.watchlistState -> {
                    deleteBtn.visibility = View.VISIBLE
                    historyBtn.visibility = View.VISIBLE
                    watchListBtn.visibility = View.GONE
                }
            }
        }

        fun assignData() {
            deleteBtn.setOnClickListener { listener.onDeleteClicked() }

            historyBtn.setOnClickListener { listener.onAddToHistoryClicked() }

            watchListBtn.setOnClickListener { listener.onAddToWatchClicked() }
        }
    }

    private inner class DescriptionViewHolder
    internal constructor(
        itemView: View,
        context: Context
    ) : RecyclerView.ViewHolder(itemView) {
        private val description: TextView = itemView.findViewById(R.id.description)
        private val more: TextView = itemView.findViewById(R.id.more)
        private val resources: Resources = context.resources

        internal fun assignData(series: Series) {
            more.setOnClickListener { showCompleteText(series) }
            description.text = getTrimmedDescription(series)
        }

        private fun setDescription(series: Series) {
            val original = series.description
            description.text = if (original == null || original.isEmpty())
                resources.getString(R.string.noDescription)
            else
                original
        }

        private fun showCompleteText(series: Series) {
            setDescription(series)
            more.visibility = View.GONE
        }

        private fun getTrimmedDescription(series: Series): String {
            val original = series.description

            if (original.isNullOrEmpty()) {
                return resources.getString(R.string.noDescription)
            }

            if (original.length <= 200) {
                more.visibility = View.GONE
                return original
            }
            more.visibility = View.VISIBLE
            return original.substring(0, 200) + "..."
        }
    }

    private inner class SeasonsListViewHolder
    internal constructor(
        itemView: View,
        private val context: Context,
        private val itemClickListener: ItemClickListener
    ) : RecyclerView.ViewHolder(itemView) {
        private val recyclerView: RecyclerView = itemView.findViewById(R.id.seasonPoster)

        internal fun assignData(series: Series?) {
            recyclerView.layoutManager = StaggeredGridLayoutManager(
                2,
                StaggeredGridLayoutManager.VERTICAL
            )
            recyclerView.itemAnimator = DefaultItemAnimator()
            recyclerView.isNestedScrollingEnabled = false

            val adapter = SeasonAdapter(context, itemClickListener, series)
            recyclerView.adapter = adapter
        }
    }
}
